lib/_stream_readable.js

```diff
@@ -669,7 +669,7 @@ Readable.prototype.on = function(ev, fn) {
       if (!state.reading) {
         process.nextTick(nReadingNextTick, this);
       } else if (state.length) {
-        emitReadable(this, state);
+        setImmediate(emitReadable.bind(null, this));
```

Comment:
`process.nextTick(emitReadable, this)` fails in one scenario, shown in parallel/test-stream3-pause-then-read's `readn(n, then)` function. At one point, one `stream.push()` is queued with `setImmediate`, and we're inside `readn(n, then)` with `n = 1234`. We're then stuck in an infinite loop: `r.read(n)` on line 46 returns `null` since `n` is higher than stream's buffer length => `r.once('readable', read)` is called which executes `process.nextTick(emitReadable, this)` inside `Readable.prototype.on` => `readable` event is emitted before the queued `stream.push()` has a chance to execute => registered `read` function executes `r.read(n)` again => infinite loop.

Queuing the `emitReadable` in a `setImmediate` lets the `stream.push()` execute, which fills the buffer (or allows for more `stream.push()` to occur which eventually fill the buffer up to `n` length).